src/Main.hs

```diff
@@ -0,0 +1,198 @@
+module Main where
+
+import qualified Network.HTTP.Simple as HTTP
+import qualified Data.Vector as Vector
+import qualified Dhall.Core as Dhall
+import qualified Dhall.Map as Map
+import qualified Data.Map.Strict as Data.Map
+import qualified Data.Set as Set
+import qualified Data.Text as Text
+import qualified Data.List as List
+import qualified Dhall.Parser as Dhall
+import qualified Data.Text.Prettyprint.Doc             as Pretty
+import qualified Data.Text.Prettyprint.Doc.Render.Text as PrettyText
+
+import Data.Bifunctor (first, second)
+import GHC.Generics (Generic)
+import Data.Maybe
+import Data.Aeson
+import Data.Text (Text)
+import Data.Set (Set)
+import Data.Map (Map)
+import Control.Applicative (optional)
+import Control.Monad (join)
+
+data BaseData = BaseData
+  { kind       :: Text
+  , apiVersion :: Text
+  } deriving (Generic, Show)
+
+instance FromJSON BaseData where
+  parseJSON = withArray "array of values" $ \arr -> withObject "baseData" (\o -> do
+    group   <- o .:? "group" .!= ""
+    kind    <- o .: "kind"
+    version <- o .: "version"
+    let apiVersion = group <> "/" <> version
+    pure BaseData{..})
+    (head $ Vector.toList arr)
+
+
+data Definition = Definition
```

Comment:
This datatype (+ its json instances) is already in http://hackage.haskell.org/package/swagger2-2.3.1.1/
perhaps we can reuse it? That would solve the 'mixing of business logic and decoding' problem without having to have multiple modules ourselves

Reply:
I didn't want to type this down so I went with `swagger2` at first, but the `swagger.json` we get from Kubernetes is broken and doesn't conform to the types there (the assumption here is that the implementation in `swagger2` is correct. The reason for the assumption is that we've been using it for years at work and it's the most correct Swagger library I've seen. On the other hand we've already seen cases in which k8s's swagger file is not the best)